### PR TITLE
Added instructions to upgrade Symfony applications to 4.x

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

### Context

Related to #24718 and lots of past questions received in Slack, Stack Overflow, etc.

The UPGRADE guides are great for individual Symfony components/bundles, but we're lacking instructions to upgrade a whole Symfony app from 2.x to 3.x or from 3.x to 4.x.

In the past we had problems with the changes introduced in https://github.com/symfony/symfony-standard which were not documented in the main UPGRADE files.

### Proposal

We could:

1) Add a new section int he UPGRADE file to explain the upgrading process for whole Symfony apps. This is what this PR proposes (but it's still just a draft).
2) We could move the changes of this PR to a new article in the Symfony Docs.
3) We could move the entire UPGRADE files to the Symfony Docs and explain the changes for individual components and whole Symfony apps.
4) ... ?

### CODE REVIEW

File: UPGRADE-4.0.md [hunk 1]
```diff
@@ -1,6 +1,40 @@
...
+
+Then, upgrade the contents of your console script and your front controller:
+
+* `bin/console`: https://github.com/symfony/recipes/blob/master/symfony/console/3.3/bin/console
+* `public/index.php`: https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/3.3/public/index.php
+
+Lastly, read the following article to add Symfony Flex to your application and
+upgrade the configuration files: https://symfony.com/doc/current/setup/flex.html
+
+If you use Symfony components instead of the whole framework, you can find below
```
Comment (Tobion) [inline UPGRADE-4.0.md]: the below upgrade instructions are also relevant if you use the whole framwork.

**Review by mickaelandrieu — APPROVED** [review]

File: UPGRADE-4.0.md [hunk 2]
```diff
@@ -1,6 +1,40 @@
...
+file and directory structure of your application:
+
+| Symfony 3.x                         | Symfony 4.x
+| ----------------------------------- | --------------------------------
+| `app/config/`                       | `config/`
+| `app/config/*.yml`                  | `config/*.yaml` and `config/packages/*.yaml`
+| `app/config/parameters.yml.dist`    | `config/services.yaml` and `.env.dist`
+| `app/config/parameters.yml`         | `config/services.yaml` and `.env`
+| `app/Resources/<BundleName>/views/` | `templates/bundles/<BundleName>/`
+| `app/Resources/`                    | `src/Resources/`
```
Comment (Taluu) [inline UPGRADE-4.0.md]: shouldn't this be `resources/` or something along those lines ? Not really source code related stuff...

Comment (javiereguiluz) [reply UPGRADE-4.0.md]: Yes ... but in the past. In old Symfony apps, we put assets, templates, translations, etc. in that app/Resources/ dir. But in new Symfony apps, assets, templates and translations have their own dir. In src/Resources/ we put "resources" related to code. For example, you could put XSD files there.

**Comment by nicolas-grekas** [comment]
We should tell about the two opt-in parameter, that should really be OK to enable most of the time (container.dumper.inline_class_loader and container.autowiring.strict_mode)

**Comment by weaverryan** [comment]
I agree with this. But shouldn’t we do it in the docs? Also (as Javier knows) there are already some Flex upgrade details on the Flex docs page (and we need to make sure those describe the best strategy, as it’s currently very laborious)

**Comment by nicolas-grekas** [comment]
@javiereguiluz what should we do here? A link to the doc?

**Comment by mickaelandrieu** [comment]
> We should tell about the two new opt-in parameters, that should really be OK to enable most of the time (container.dumper.inline_class_loader and container.autowiring.strict_mode)

Can you explain in which cases something can not be OK? PrestaShop still have his own autoloader for instance, and I have the feeling that we wont be able to enjoy all the perfs improvements you have done in 3.4 ^^

**Comment by nicolas-grekas** [comment]
@mickaelandrieu when the autoloader does something more than just finding + requiring a file. IE runs some logic before or after the load.

**Comment by stof** [comment]
@mickaelandrieu the behavior of `container.autowiring.strict_mode` is that it disables the BC layer for the SF 3.2 autowiring behavior. If you don't get any autowiring-related deprecation warning in 3.3+, you can turn on this parameter (or you can turn it on already, and then fix any autowiring error you get).

For `container.dumper.inline_class_loader`, there are 2 cases breaking it:

- crazy autoloading logic (cf @nicolas-grekas's comment). Note that Swiftmailer is a case doing that.
- using `$kernel->loadClassCache()` (which is deprecated, and actually *hurts* performance on PHP 7+), so this should just be removed from the project.

**Comment by mickaelandrieu** [comment]
> when the autoloader does something more than just finding + requiring a file.

:'( fml

Thank for information, both of you :+1:

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Pull request merged**